### DISCUSSION TITLE
Stuff

### DISCUSSION THREAD

**Discussion by gateway** [rkhamilton/vqgan-clip-generator#4] [Ideas]
Btw thank you so much for taking the time to do this. I have been playing around with the prior code.  I'm surprised I got this to run on windows as well (my Titan RTX is on windows).

Would you mind enabling the discussion tab in github for this?

Oh I did notice one issue on windows.. the output filename blows if you have things like ^ or | type of separators in your text prompt.

One question about video.. can you add a way to set a downscale size.. I have a bunch of 4k footage I wanted to test this on a downscale option when extracting the images from the video would be awesome.

I think the VF option in ffmpeg works.. ```-vf scale=192:168``` 

Thoughts?

**Comment by rkhamilton** [comment]
Thanks for the note! I've just put this all together in the last week.

I'll enable discussions shortly. I've never had anybody want to talk about
any of my code before so I've never even looked at is. I assumed it was
enabled!

I also noticed that the filename breaks if you have phrase separators. In
my examples I just manually enter filenames instead of sanitizing them. The
simplest use case was just manual filenames, but I was always wanting to
rename them to the text prompt. I'll think about whether I should break out
some kind of "first safe part of the text prompt" method for naming files,
or just rely on the user to use safe filenames manually.

I want to warn you that I thought I was done with it and called it v 1.0,
but a friend of mine suggested I incorporate ESRGAN upscaling, which
excited me and I'm close to pushing a 1.1 that breaks compatibility with
most stuff. My generate.*_video methods all wrapped up generating new
images with encoding video to make for a compact structure, but that made
it hard to easily integrate ESRGAN upscaling. The new syntax will look like
generate video frames
upscale video frames
encode video.

Re: downscaling video, I'll take a look. It should be easy enough. In your
example is it just generating files that take up too much space, or is it
failing to run?


On Sun, Oct 17, 2021 at 6:28 PM Gateway ***@***.***> wrote:

> Btw thank you so much for taking the time to do this. I have been playing
> around with the prior code. I'm surprised I got this to run on windows as
> well (my Titan RTX is on windows).
>
> Would you mind enabling the discussion tab in github for this?
>
> Oh I did notice one issue on windows.. the output filename blows if you
> have things like ^ or | type of separators in your text prompt.
>
> One question about video.. can you add a way to set a downscale size.. I
> have a bunch of 4k footage I wanted to test this on a downscale option when
> extracting the images from the video would be awesome.
>
> I think the VF option in ffmpeg works.. -vf scale=192:168
>
> Thoughts?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/rkhamilton/vqgan-clip-generator/issues/1>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACAEKSBY4AXFNGU2BS5PS63UHNEXZANCNFSM5GFJPNFQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Ryan

**Comment by gateway** [comment]
I started a new discussion about this.. didn't realize this was moved to a discussion.

I had an idea though while thinking about you wanting to keep the text as a filename.. it would be great if we stored the text and settings in the metadata of the image.  Exif data can be stored in the user-comment field.  So if you wanted to go back to an image to see what settings you had you could just read the meta data.. thoughts? :)

**Reply by rkhamilton** [reply]
It's a good idea, and I'm already doing that for video files. The video_tools.encode_video() method has a metadata argument. In my video examples that is used to store the original text prompts. 

I'm not currently storing metadata in images. Other versions of VQGAN+CLIP out there actually do that, but I wasn't able to get the metadata packages to install correctly on my PC and just sort of gave up and moved on. I'd like to add it back for images. I'll keep it in mind for the future. It's certainly nice to have.

**Reply by gateway** [reply]
I think PIL has the ability to write to the user comment section..

```from PIL import Image
im = Image.open("Tests/images/hopper.jpg")
exif = im.getexif()
exif[0x9286] = "test"
im.save("out.jpg", exif=exif)```

exif tags..

```0x9286 | 37510 | Photo | Exif.Photo.UserComment | Comment | A tag for Exif users to write keywords or comments on the image besides those in <ImageDescription>, and without the character code limitations of the <ImageDescription> tag```

**Reply by rkhamilton** [reply]
I just tried that, plus PIL.PngImagePlugin, and I'm not seeing any data show up where I can read it as an exif field with Lightroom. Both methods don't throw an error so they are probably writing something somewhere, but I don't see that it's in a usefully visible place. If you can develop a working code snippet that applies to a png output from vqgan_clip, and the data shows up in a way that you can usefully read it, I'd be happy to incorporate it.

**Reply by gateway** [reply]
hmm further digging, it seems that PNG had some sort of EXIF data attached to it but not the normal type we would see in a dng, raw or jpg.  You can add key -> value chunks in such as "text" -> "a picture in the style of van Gogh" , "config options" -> "etc" 

Unfortunately not a lot of apps read the chunk data.  I wrote some stuff to the png file then used png check or imagemagick identify command to pull out that data gain.

```identify -verbose NewPath.png| grep "MyNewString"```

```gateway@iMac Desktop % pngcheck -c -v -t NewPath.png 
File: NewPath.png (1208007 bytes)
  chunk IHDR at offset 0x0000c, length 13
    896 x 896 image, 24-bit RGB, non-interlaced
  chunk tEXt at offset 0x00025, length 20, keyword: MyNewString
    A string
  chunk tEXt at offset 0x00045, length 13, keyword: MyNewInt
    1234
  chunk IDAT at offset 0x0005e, length 65536
    zlib: deflated, 32K window, default compression
  chunk IDAT at offset 0x1006a, length 65536
  chunk IDAT at offset 0x20076, length 65536
  chunk IDAT at offset 0x30082, length 65536
  chunk IDAT at offset 0x4008e, length 65536
  chunk IDAT at offset 0x5009a, length 65536
  chunk IDAT at offset 0x600a6, length 65536
  chunk IDAT at offset 0x700b2, length 65536
  chunk IDAT at offset 0x800be, length 65536
  chunk IDAT at offset 0x900ca, length 65536
  chunk IDAT at offset 0xa00d6, length 65536
  chunk IDAT at offset 0xb00e2, length 65536
  chunk IDAT at offset 0xc00ee, length 65536
  chunk IDAT at offset 0xd00fa, length 65536
  chunk IDAT at offset 0xe0106, length 65536
  chunk IDAT at offset 0xf0112, length 65536
  chunk IDAT at offset 0x10011e, length 65536
  chunk IDAT at offset 0x11012a, length 65536
  chunk IDAT at offset 0x120136, length 28029
  chunk IEND at offset 0x126ebf, length 0```

so I you uncomment your line info.add.text here https://github.com/rkhamilton/vqgan-clip-generator/blob/4bcc8cc01be4587a7ad0ec01cef9ec02fad051ef/src/vqgan_clip/engine.py#L166 it should be readable by those programs.. of course not really cool EXIF data but still data thats stored.

**Reply by rkhamilton** [reply]
I'll stew on it for a while. Anything we add here wouldn't survive Real-ESRGAN upscaling unless I build in a metadata transfer process, which doesn't excite me. I struggled with a bigger xmp/exif/IPTC project a couple years ago that turned out to be a big pain in the butt to deal with and I'm not looking forward to digging it back up until I can solve it for real.

**Reply by gateway** [reply]
So @rkhamilton I thought of a way to use the meta data in png files.  I'm doing tests right now with things like

"A nike made out of liquid metal" with changes to iteration value and save every.. so I might try i = 10, save 2, then i = 20 , save 5 and just start to play around with what Im starting to like ideally ill write a script to change these values and file name with the text... HOWEVER.. if the text , iterations / save every and others where meta chunks (key -> value) in the png files I could use imagemagick montage function (takes a bunch of images and creates this cool montage of images, you can also add text under it or to the side.. so I could have my test of various iterations based upon text but i could visually look at what values I put in.

I think you already have the text set up but its uncommitted in the png save function, adding a few would be awesome.. it wouldn't heart anyone but anyone wanting to create some vary unique montages or hell have a script on a web server that creates a page and uses the metadata to populate text, iterations blah blah blah? :)

**Reply by rkhamilton** [reply]
Ok, it's implemented for v1.2. Separate chunks for each relevant parameter.
![image](https://user-images.githubusercontent.com/8406344/138380458-efd09c23-5d8f-433d-9aa0-a82e8c2664d5.png)

**Reply by gateway** [reply]
Dude your awesome thank you so much for doing this an sometimes entertaining my crazy thoughts and ideas..

**Reply by rkhamilton** [reply]
Once you get a hold of 1.2 and generate your method for making annotated mosaics please share it with me. I was just scanning folders of thumbnail images to evaluate parameters.

**Reply by gateway** [reply]
sure thing.. this was done a while ago .. of course this was sooooo many different prompts from a semi random text generator.  https://imgur.com/gallery/KIZuRyJ

**Reply by rkhamilton** [reply]
I was inspired by adding metadata to these images, so last night I worked up changes to add metadata to video as well.
title = text_prompts
comments = a string of generation parameter
example: iterations: 15, image_prompts: [], noise_prompts: [], init_weight_method: decay, init_weight 1.00, init_image video_frames, change_prompt_every 0, seed 796212828143400

In the next version, which I should push today once some automated tests finish.
